File: Source/MemoryReader.cpp
#include "MemoryReader.h"
#include <winioctl.h>
#include <vector>

#define IOCTL_READ_MEMORY CTL_CODE(FILE_DEVICE_UNKNOWN, 0x800, METHOD_BUFFERED, FILE_ANY_ACCESS)
#define IOCTL_WRITE_MEMORY CTL_CODE(FILE_DEVICE_UNKNOWN, 0x801, METHOD_BUFFERED, FILE_ANY_ACCESS)

namespace prt {

#pragma pack(push, 1)
struct READ_MEM_REQ {
    uint32_t ProcessId;
    uint64_t TargetAddress;
    uint64_t Size;
};

struct WRITE_MEM_REQ {
    uint32_t ProcessId;
    uint64_t TargetAddress;
    uint64_t Size;
};
#pragma pack(pop)

MemoryReader::MemoryReader()
    : _hProcess(GetCurrentProcess()),
      _baseAddress((uintptr_t)GetModuleHandle(NULL)),
      _pid(GetCurrentProcessId()),
      _hDriver(INVALID_HANDLE_VALUE) {}

MemoryReader::MemoryReader(HANDLE hProcess, uintptr_t baseAddress, uint32_t pid, HANDLE hDriver)
    : _hProcess(hProcess), _baseAddress(baseAddress), _pid(pid), _hDriver(hDriver) {}

MemoryReader::~MemoryReader() {}

bool MemoryReader::ReadBytes(uintptr_t address, void* buffer, size_t size) {
    if (_hDriver != INVALID_HANDLE_VALUE) {
        READ_MEM_REQ req = { _pid, (uint64_t)address, (uint64_t)size };
        DWORD bytesReturned = 0;
        return DeviceIoControl(_hDriver, IOCTL_READ_MEMORY, &req, sizeof(req), buffer, (DWORD)size, &bytesReturned, NULL);
    }
    SIZE_T bytesRead = 0;
    return ReadProcessMemory(_hProcess, (LPCVOID)address, buffer, size, &bytesRead) && bytesRead == size;
}

uint32_t MemoryReader::ReadUInt32(uintptr_t address) {
    return Read<uint32_t>(address);
}

int32_t MemoryReader::ReadInt32(uintptr_t address) {
    return Read<int32_t>(address);
}

float MemoryReader::ReadFloat(uintptr_t address) {
    return Read<float>(address);
}

uintptr_t MemoryReader::ReadPointer(uintptr_t address, int offset) {
    return (uintptr_t)Read<uint32_t>(address + offset);
}

uintptr_t MemoryReader::ResolvePointerChain(uintptr_t baseAddr, const std::vector<int>& offsets, int finalOffset) {
    uintptr_t addr = baseAddr;
    for (int offset : offsets) {
        addr = ReadPointer(addr + offset);
        if (addr == 0) return 0;
    }
    return addr + finalOffset;
}

std::string MemoryReader::ReadString(uintptr_t address, size_t maxLength) {
    std::vector<char> buf(maxLength);
    if (!ReadBytes(address, buf.data(), maxLength)) return "";
    buf.push_back('\0');
    return std::string(buf.data());
}

bool MemoryReader::WriteBytes(uintptr_t address, const void* buffer, size_t size) {
    if (_hDriver != INVALID_HANDLE_VALUE) {
        std::vector<uint8_t> ioBuffer(sizeof(WRITE_MEM_REQ) + size);
        auto* req = reinterpret_cast<WRITE_MEM_REQ*>(ioBuffer.data());
        req->ProcessId = _pid;
        req->TargetAddress = (uint64_t)address;
        req->Size = (uint64_t)size;
        memcpy(ioBuffer.data() + sizeof(WRITE_MEM_REQ), buffer, size);

        DWORD bytesReturned = 0;
        return DeviceIoControl(_hDriver, IOCTL_WRITE_MEMORY, ioBuffer.data(), (DWORD)ioBuffer.size(), NULL, 0, &bytesReturned, NULL);
    }
    SIZE_T bytesWritten = 0;
    return WriteProcessMemory(_hProcess, (LPVOID)address, buffer, size, &bytesWritten) && bytesWritten == size;
}

} // namespace prt

File: 4RTools Adjusted/4RTools-main/Utils/MuhBotCore/MemoryReader.cs
#nullable enable
using System;
using System.Runtime.InteropServices;
using System.Text;

namespace _4RTools.Utils.MuhBotCore;

/// <summary>
/// Typed read/write and pointer chain resolution for process memory.
/// Supports direct ReadProcessMemory and optional driver-backed reads.
/// </summary>
public class MemoryReader
{
    private readonly ProcessManager _processManager;

    public MemoryReader(ProcessManager processManager)
    {
        _processManager = processManager;
    }

    private IntPtr Handle => _processManager.Handle;
    private IntPtr Base => _processManager.BaseAddress;
    private IntPtr WHandle => _processManager.WriteHandle;

    /// <summary>True when we have a valid process handle (attached).</summary>
    public bool IsValid => _processManager.IsAttached;

    public void CloseWriteHandle() => _processManager.CloseWriteHandle();

    public bool ReadBytes(IntPtr address, byte[] buffer)
    {
        return InternalRead(address, buffer, (uint)buffer.Length);
    }

    public bool WriteBytes(IntPtr address, byte[] buffer)
    {
        return InternalWrite(address, buffer, (uint)buffer.Length);
    }

    public uint ReadUInt32(IntPtr address)
    {
        var buf = new byte[4];
        if (!InternalRead(address, buf, 4))
            return 0;
        return BitConverter.ToUInt32(buf, 0);
    }

    public int ReadInt32(IntPtr address)
    {
        var buf = new byte[4];
        if (!InternalRead(address, buf, 4))
            return 0;
        return BitConverter.ToInt32(buf, 0);
    }

    public float ReadFloat(IntPtr address)
    {
        var buf = new byte[4];
        if (!InternalRead(address, buf, 4))
            return 0f;
        return BitConverter.ToSingle(buf, 0);
    }

    public IntPtr ReadPointer(IntPtr address, int offset = 0)
    {
        var buf = new byte[4];
        if (!InternalRead(IntPtr.Add(address, offset), buf, 4))
            return IntPtr.Zero;

        uint raw = BitConverter.ToUInt32(buf, 0);
        if (raw == 0) return IntPtr.Zero;
        return IntPtr.Size == 8
            ? new IntPtr(unchecked((long)raw))
            : new IntPtr(unchecked((int)raw));
    }

    /// <summary>
    /// Resolve multi-level pointer chain: base + offsets[0] -> read ptr -> + offsets[1] -> ... -> final + lastOffset.
    /// </summary>
    public IntPtr ResolvePointerChain(IntPtr baseAddr, int[] offsets, int finalOffset = 0)
    {
        IntPtr addr = baseAddr;
        for (int i = 0; i < offsets.Length; i++)
        {
            addr = IntPtr.Add(addr, offsets[i]);
            addr = ReadPointer(addr);
            if (addr == IntPtr.Zero)
                return IntPtr.Zero;
        }
        return IntPtr.Add(addr, finalOffset);
    }

    /// <summary>
    /// Resolve from process base: baseAddress + firstOffset, then follow pointer chain.
    /// </summary>
    public IntPtr ResolvePointerChainFromBase(int[] offsets, int finalOffset = 0)
    {
        IntPtr addr = ReadPointer(IntPtr.Add(Base, offsets[0]));
        if (addr == IntPtr.Zero)
            return IntPtr.Zero;
        for (int i = 1; i < offsets.Length; i++)
        {
            addr = ReadPointer(IntPtr.Add(addr, offsets[i]));
            if (addr == IntPtr.Zero)
                return IntPtr.Zero;
        }
        return IntPtr.Add(addr, finalOffset);
    }

    public string ReadString(IntPtr address, int maxLength = 64, Encoding? encoding = null)
    {
        encoding ??= Encoding.ASCII;
        var buf = new byte[maxLength];
        if (!InternalRead(address, buf, (uint)maxLength))
            return string.Empty;
        int end = Array.IndexOf(buf, (byte)0);
        if (end < 0) end = maxLength;
        return encoding.GetString(buf, 0, end).TrimEnd('\0');
    }

    public bool WriteUInt32(IntPtr address, uint value)
    {
        var buf = BitConverter.GetBytes(value);
        return InternalWrite(address, buf, (uint)buf.Length);
    }

    public bool WriteInt32(IntPtr address, int value)
    {
        var buf = BitConverter.GetBytes(value);
        return InternalWrite(address, buf, (uint)buf.Length);
    }

    public bool WriteFloat(IntPtr address, float value)
    {
        var buf = BitConverter.GetBytes(value);
        return InternalWrite(address, buf, (uint)buf.Length);
    }

    public IntPtr BaseAddress => Base;

    private bool InternalRead(IntPtr address, byte[] buffer, uint size)
    {
        IntPtr driverHandle = _processManager.DriverHandle;
        if (driverHandle != IntPtr.Zero && driverHandle.ToInt64() != -1)
        {
            var req = new Native.READ_MEM_REQ
            {
                ProcessId = (uint)_processManager.ProcessId,
                TargetAddress = (ulong)address.ToInt64(),
                Size = size
            };

            return Native.DeviceIoControl(
                driverHandle,
                Native.IOCTL_READ_MEMORY,
                ref req,
                (uint)Marshal.SizeOf<Native.READ_MEM_REQ>(),
                buffer,
                size,
                out _,
                IntPtr.Zero);
        }

        return Native.ReadProcessMemory(Handle, address, buffer, size, out _);
    }

    private bool InternalWrite(IntPtr address, byte[] data, uint size)
    {
        IntPtr driverHandle = _processManager.DriverHandle;
        if (driverHandle != IntPtr.Zero && driverHandle.ToInt64() != -1)
        {
            // Build combined buffer: [WRITE_MEM_REQ header][data]
            int headerSize = Marshal.SizeOf<Native.WRITE_MEM_REQ>();
            var req = new Native.WRITE_MEM_REQ
            {
                ProcessId = (uint)_processManager.ProcessId,
                TargetAddress = (ulong)address.ToInt64(),
                Size = size
            };

            byte[] inputBuffer = new byte[headerSize + size];
            // Marshal the header struct into the buffer
            IntPtr headerPtr = Marshal.AllocHGlobal(headerSize);
            try
            {
                Marshal.StructureToPtr(req, headerPtr, false);
                Marshal.Copy(headerPtr, inputBuffer, 0, headerSize);
            }
            finally
            {
                Marshal.FreeHGlobal(headerPtr);
            }
            // Copy data payload after the header
            Array.Copy(data, 0, inputBuffer, headerSize, size);

            return Native.DeviceIoControl(
                driverHandle,
                Native.IOCTL_WRITE_MEMORY,
                inputBuffer,
                (uint)inputBuffer.Length,
                Array.Empty<byte>(),
                0,
                out _,
                IntPtr.Zero);
        }

        return Native.WriteProcessMemory(WHandle, address, data, size, out _);
    }
}
